sdk/core/core/src/az_log.c

```diff
@@ -27,38 +27,61 @@ void az_log_set_callback(az_log_message_fn az_log_message_callback)
   _az_log_message_callback = az_log_message_callback;
 }
 
-void _az_log_write(az_log_classification classification, az_span message)
+// _az_log_write_engine is a function private to this .c file; it contains the code to handle
+// _az_log_should_write & _az_log_write.
+//
+// If log_it is false, then the function returns true or false indicating whether the message
+// should be logged (without actually logging it). if log_it is false, then the function returns
+// true or false indicating whether the message should be logged (without actually logging it).
+//
+// If log_it is true, then the function logs the message (if it should) and returns true or
+// false indicating whether it was logged.
+static bool _az_log_write_engine(bool log_it, az_log_classification classification, az_span message)
 {
-  if (_az_log_message_callback != NULL && _az_log_should_write(classification))
-  {
-    _az_log_message_callback(classification, message);
-  }
-}
+  // Copy the volatile fields to local variables so that they don't change within this function
+  az_log_message_fn const callback = _az_log_message_callback;
+  az_log_classification const* const classifications = _az_log_classifications;
 
-bool _az_log_should_write(az_log_classification classification)
-{
-  if (_az_log_message_callback == NULL)
+  if (callback == NULL)
   {
     // If no one is listening, don't attempt to log.
     return false;
   }
 
-  if (_az_log_classifications == NULL)
+  // If the user hasn't registered any classifications, then we log everything.
+  bool const log_everything = (classifications == NULL);
```

Comment:
What did you do here?
If log_everything is true, then when does callback get invoked?
Did my code not work? This code is MUCH harder to understand.

Reply:
if log_everything is true, it goes to case true, i.e. jumps right inside the loop.

Reply:
Here is with the single-item array, as you suggested: https://github.com/Azure/azure-sdk-for-c/pull/697

Reply:
Another way - do everything inside that if(): https://github.com/Azure/azure-sdk-for-c/pull/698

Reply:
I'll apply any the one you pick, or if you suggest something else I can write/add it.